### DISCUSSION TITLE
How to make custom error messages more maintainable?

### DISCUSSION THREAD

**Discussion by matjaeck** [sinclairzx81/typebox#872] [General]
Hi,
it seems that the suggested approach to handle custom errors is to use SetErrorFunction and look for the presence of a `errorMessage` property in the schema params like this:

```ts
SetErrorFunction(param => `errorMessage` in param.schema ? param.schema.errorMessage : DefaultErrorFunction(param));
```
And then define `errorMessage` as part of the schema options like so:
```ts
const schema = Object({
   test: Type.String({ minLength: 5, errorMessage: "Foo" }),
});
```
But `errorMessage` is an unknown property to SchemaOptions, it will happily accept things like `errorMessagez` and then fallback to the default messages.

How can we make using a custom `errorMessage` property more maintainable? How have other users solved this?

- Change SetErrorFunction to throw when `errorMessage` is not present and build tests around it?
- Create a wrapper function that accepts the schema type, the default parameters and merges them with a errorMessage parameter?
- Change the type of TSchema to accept an optional `errorMessage` property?

Or am I missing something?

**Comment by sinclairzx81** [comment]
@matjaeck Hi,

> But errorMessage is an unknown property to SchemaOptions, it will happily accept things like errorMessagez and then fallback to the default messages.

Unfortunately, there's not really an elegant way to solve for misspelt properties (this is because SchemaOptions is an open ended interface type that can be assigned any additional properties). It is however possible to use interface/module augmentation to add additional known properties to the TB option interfaces. This will give auto-complete for additional "known" properties as well as static checking.

[TypeScript Link Here](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxnAL5wBmUEIcA5AAIDOwAdiQDYHBQD0MdBgBGEAB4MAUFP784ASXjAWcSCzYiuGODAhwCAVwDmhkBnbxkAITgcYGKFQIkMqjroAWOqhC5cIAHcOYzhAglQ4FkMSTy8iOFQIQzhjDHhZeSM9AFoSOjBtBwN2ABMowmIyWIwSAGsQ6mgDUtLgYgh2Ai5GOvYg9gY1WkxYYDccEUNlZRZMuH74dU1tOAAKAmdgOoBKOBr6pqgWto6unoZDdj6BobARx2I3ADpnqVLanigdEAhSw1WzDYnB4fEEwjEkkQUjgcAw4kgsDslkczlceBqIAIAHkwGdVAgYbC4VBaFAALJuFgENIAfgAXBUoCEieQpGypGhMM98Mz2MY1khHGTKRoaRhGc4uCwdOQ9rC5HAboF2MMIKMhHAAGRhWj83TCGRcjA8mB8gVC0nQUXUtKMhgQOL9IZyuCK5Wq+7qx6RHX5Um1eBCTBG4Sm82Ckkiqk0gBe9sdCwgLr2iquHrVGt9cHT-RVBswQA)

```typescript
import { Type, Static } from '@sinclair/typebox'

// It is possible to augument TB interfaces in the following way such that you get 
// auto-complete and static checking for additional 'known' properties, but it is
// not possible (afaik) check for additional 'unknown' properties..
declare module '@sinclair/typebox' {
  export interface SchemaOptions {
    errorMessage?: string
  }
}

Type.String({ errorMessage: false })   // known property & wrong type

Type.String({ errorMessage: 'oh no' }) // known property & correct type

Type.String({ errorMessagz: 'oh no' }) // unknown property & unknown type
```



Hope this helps
S

**Reply by matjaeck** [reply]
Thank you for pointing me in the right direction and taking the time to write it down.
